File: Python/295__Find_Median_from_Data_Stream.py
from heapq import heappush, heappop, heapreplace, heapify

class MedianFinder(object):

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.maxHeap = []
        self.minHeap = []


    def addNum(self, num):
        """
        :type num: int
        :rtype: void
        """
        heappush(self.maxHeap, -num)
        heappush(self.minHeap, -heappop(self.maxHeap))
        if len(self.minHeap) > len(self.maxHeap):
            heappush(self.maxHeap, -heappop(self.minHeap))


    def findMedian(self):
        """
        :rtype: float
        """
        if len(self.maxHeap) > len(self.minHeap):
            return float(-self.maxHeap[0])

        return ((-self.maxHeap[0] + self.minHeap[0] + 0.00 )/2)




# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()

File: Java/295__Find_Median_from_Data_Stream.java
class MedianFinder {

    PriorityQueue<Integer> min;
    PriorityQueue<Integer> max;

    /** initialize your data structure here. */
    public MedianFinder() {
        min = new PriorityQueue<>();
        max = new PriorityQueue<>(Collections.reverseOrder());
    }

    public void addNum(int num) {
        if (!min.isEmpty() && num > min.peek())
            min.offer(num);
        else
            max.offer(num);
        if (max.size() - min.size() > 1 || max.size() < min.size())
            balance(max, min);
    }

    public double findMedian() {
        if ((max.size() + min.size()) % 2 == 0) {
            return ((double) max.peek() + (double) min.peek()) / 2.0;
        } else {
            return (double) max.peek();
        }
    }

    private void balance(PriorityQueue<Integer> max, PriorityQueue<Integer> min) {
        while (max.size() - min.size() > 1) {
            min.offer(max.poll());
        }
        while (max.size() < min.size()) {
            max.offer(min.poll());
        }
    }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */
